### PR TITLE
 [Graphbolt] Define API for Neighbor sampling

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Define API interface for graphbolt sampling.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 1]
```diff
@@ -115,6 +115,17 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
   c10::intrusive_ptr<SampledSubgraph> InSubgraph(
       const torch::Tensor& nodes) const;
 
+  /**
+   * @brief Sample neighbors for a given set of nodes.
```
Comment (frozenbugs) [inline graphbolt/include/graphbolt/csc_sampling_graph.h:119]: Sample neighboring edges of the given nodes and return the induced subgraph.

**Review by frozenbugs — APPROVED** [review]

File: tests/python/pytorch/graphbolt/test_csc_sampling_graph.py [hunk 2]
```diff
@@ -385,6 +385,42 @@ def test_in_subgraph_heterogeneous():
...
+    num_edges = 12
+    indptr = torch.LongTensor([0, 3, 5, 7, 9, 12])
+    indices = torch.LongTensor([0, 1, 4, 2, 3, 0, 1, 1, 2, 0, 3, 4])
+    assert indptr[-1] == num_edges
+    assert indptr[-1] == len(indices)
+
+    # Construct CSCSamplingGraph.
+    graph = gb.from_csc(indptr, indices)
+
+    # Extract in subgraph.
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_csc_sampling_graph.py:411]: `# Generate subgraph via sample neighbors.

File: graphbolt/include/graphbolt/csc_sampling_graph.h (hunk 1 above)
Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h]: Fixed

File: tests/python/pytorch/graphbolt/test_csc_sampling_graph.py (hunk 2 above)
Comment (peizhou001) [reply tests/python/pytorch/graphbolt/test_csc_sampling_graph.py:411]: Fixed

File: graphbolt/include/graphbolt/csc_sampling_graph.h [hunk 3]
```diff
@@ -115,6 +115,18 @@ class CSCSamplingGraph : public torch::CustomClassHolder {
...
+   * @brief Sample neighboring edges of the given nodes and return the induced
+   * subgraph.
+   *
+   * @param nodes The nodes from which to sample neighbors.
+   *
+   * @return An intrusive pointer to a SampledSubgraph object containing the
+   * sampled graph's information.
+   */
+  c10::intrusive_ptr<SampledSubgraph> SampleNeighbors(
+      const torch::Tensor& nodes) const;
```
Comment (Rhett-Ying) [inline graphbolt/include/graphbolt/csc_sampling_graph.h:128]: This is just for PLACE_HOLDER as parameters like `fanout` is missing?

File: tests/python/pytorch/graphbolt/test_csc_sampling_graph.py [hunk 4]
```diff
@@ -385,6 +385,42 @@ def test_in_subgraph_heterogeneous():
...
+    assert indptr[-1] == len(indices)
+
+    # Construct CSCSamplingGraph.
+    graph = gb.from_csc(indptr, indices)
+
+    # Generate subgraph via sample neighbors.
+    nodes = torch.LongTensor([1, 3, 4])
+    subgraph = graph.sample_neighbors(nodes)
+
+    # Verify in subgraph.
```
Comment (Rhett-Ying) [inline tests/python/pytorch/graphbolt/test_csc_sampling_graph.py:415]: remove `in`?

File: graphbolt/include/graphbolt/csc_sampling_graph.h (hunk 3 above)
Comment (peizhou001) [reply graphbolt/include/graphbolt/csc_sampling_graph.h:128]: Yes

File: tests/python/pytorch/graphbolt/test_csc_sampling_graph.py (hunk 4 above)
Comment (peizhou001) [reply tests/python/pytorch/graphbolt/test_csc_sampling_graph.py:415]: Maybe pk, beacuse it means information `in`  subgraph

**Pull request merged**